### PR TITLE
ML themed usecase

### PR DESCRIPTION
**Pull request by adswa**
based on the dvc comparison.

TODO
- [x] code cast for talk
- [x] reference in dvc comparison

### CODE REVIEW

**Comment by adswa** [comment]
Ping @mih. Here's a usecase for the ML-themed workshop. Would you mind skimming through the approach and check whether its okay for you or if you can think of improvements or other useful things to showcase? :) We should also write out an email about this.

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 1]
```diff
@@ -0,0 +1,840 @@
...
+
+   datalad containers-run -n software \
+     -m "Prepare the data for categories golf balls and parachutes" \
+     --input 'data/raw/imagenette2-160/train/n03445777' \
+     --input 'data/raw/imagenette2-160/val/n03445777' \
+     --input 'data/raw/imagenette2-160/train/n03888257' \
+     --input 'data/raw/imagenette2-160/val/n03888257' \
+     --output 'data/train.csv' \
+     --output 'data/test.csv' \
+     "python3 code/prepare.py"
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:574]: Download can be sped up with

`datalad -c datalad.runtime.max-annex-jobs=5 containers-run ...`

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+
+
+And now, we can execute the scripts.
+First, train the model::
+
+   datalad containers-run -n software \
+     -m "Train an SGD classifier on the data" \
+     --input 'data/raw/imagenette2-160/train/*' \
+     --output 'model.joblib' \
+     "python3 code/train.py"
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:688]: Should this be the same input spec as above, otherwise pulls data that we made no labels for.

Comment (mih) [reply docs/code_from_chapters/usecase_ml_code.rst:688]: I see:

```
[INFO   ] == Command start (output follows) ===== 
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_stochastic_gradient.py:573: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO   ] == Command exit (modification check follows) ===== 
```

maybe the code should be tuned.

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+
+
+Then, evaluate performance::
+
+   datalad containers-run -n software \
+     -m "Evaluate SGD classifier on test data" \
+     --input 'data/raw/imagenette2-160/val/*' \
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:695]: Same as above, pulls more than needed.

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+     --output 'accuracy.json' \
+     "python3 code/evaluate.py"
+
+Repeat after tuning!
+""""""""""""""""""""
+
+We can now demonstrate how the run records come in handy when we change an aspect of the analysis.
+Let's increase the number of iterations turing training from 10 to 100 (here done via the stream editor :term:`sed`)::
+
+   sed -i 's/SGDClassifier(max_iter=10)/SGDClassifier(max_iter=100)/g' code/train.py
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:705]: Ah, I see...it was intentional ....

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 5]
```diff
@@ -0,0 +1,840 @@
...
+
+
+Then, evaluate performance::
+
+   datalad containers-run -n software \
+     -m "Evaluate SGD classifier on test data" \
+     --input 'data/raw/imagenette2-160/val/*' \
+     --output 'accuracy.json' \
+     "python3 code/evaluate.py"
+
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:698]: At this point I have:

```
add(ok): /tmp/ml/ml-project/code/__pycache__/train.cpython-37.pyc (file)
```

`__pycache__` should be in `.gitignore`

Comment (mih) [reply docs/code_from_chapters/usecase_ml_code.rst:698]: ~~Also `model.joblib` get's annexed. Maybe this should be treated as a temp-file and ignored too? Didn't check the code of needs to be saved....~~

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+       train_csv_path = repo_path / "data/train.csv"
+       train_data, labels = load_data(train_csv_path)
+       rf = RandomForestClassifier()
+       trained_model = rf.fit(train_data, labels)
+       dump(trained_model, repo_path / "model.joblib")
+
+   if __name__ == "__main__":
+       repo_path = Path(__file__).parent.parent
+       main(repo_path)
+   EOT
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:767]: could be tweaked into a more elaborate `sed` call too

File: docs/code_from_chapters/usecase_ml_code.rst [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+
+Let's save it::
+
+   datalad save -m "Add SGD classification script" code/train.py
+
+The next script loads the trained classifier from disk and evaluates it on the validation data.
+To evaluate the model performance, it calculates the accuracy of the prediction, i.e., the proportion of correctly labeled images, prints it to the terminal, and saves it into a json file in the superdataset::
```
Comment (mih) [inline docs/code_from_chapters/usecase_ml_code.rst:643]: JSON file

**Review by mih — APPROVED** [review]
This is cool stuff. Simple yet comprehensive. I like it.

File: docs/code_from_chapters/usecase_ml_code.rst (hunk 2 above)
Comment (adswa) [reply docs/code_from_chapters/usecase_ml_code.rst:688]: (the next script change tunes it, increases number of iterations)

File: docs/code_from_chapters/usecase_ml_code.rst (hunk 7 above)
Comment (adswa) [reply docs/code_from_chapters/usecase_ml_code.rst:643]: done

File: docs/code_from_chapters/usecase_ml_code.rst (hunk 3 above)
Comment (adswa) [reply docs/code_from_chapters/usecase_ml_code.rst:695]: thanks for noticing!

File: docs/code_from_chapters/usecase_ml_code.rst (hunk 1 above)
Comment (adswa) [reply docs/code_from_chapters/usecase_ml_code.rst:574]: thx

**Pull request merged**